RecoJets/JetProducers/plugins/FixedGridRhoProducerFastjetFromRecHit.cc

```diff
@@ -0,0 +1,180 @@
+  desc.add<edm::InputTag>("ebRecHitsTag", edm::InputTag("hltEcalRecHit", "EcalRecHitsEB"));
+  desc.add<edm::InputTag>("eeRecHitsTag", edm::InputTag("hltEcalRecHit", "EcalRecHitsEE"));
+  desc.add<bool>("skipHCAL", false);
+  desc.add<bool>("skipECAL", false);
+  //eThresHB/HE are from RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHBHE_cfi.py
+  desc.add<std::vector<double> >("eThresHB", {0.1, 0.2, 0.3, 0.3});
+  desc.add<std::vector<double> >("eThresHE", {0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2});
+  desc.add<double>("maxRapidity", 2.5);
+  desc.add<double>("gridSpacing", 0.55);
+  descriptions.add("hltFixedGridRhoProducerFastjetFromRecHit", desc);
+}
+
+FixedGridRhoProducerFastjetFromRecHit::~FixedGridRhoProducerFastjetFromRecHit() {}
+
+void FixedGridRhoProducerFastjetFromRecHit::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  std::vector<fastjet::PseudoJet> inputs;
+  auto const &thresholds = iSetup.getData(ecalPFRechitThresholdsToken_);
+
+  if (skipHCAL_ && skipECAL_) {
+    throw cms::Exception("FixedGridRhoProducerFastjetFromRecHit")
+        << "skipHCAL and skipECAL both can't be True. Make at least one of them False.";
+  }
+
+  if (!skipHCAL_) {
+    for (const auto &hit : iEvent.get(hbheRecHitsTag_)) {
+      if (passedHcalNoiseCut(hit)) {
+        TLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, iSetup.getData(caloGeometryToken_));
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+  }
+
+  if (!skipECAL_) {
+    for (const auto &hit : iEvent.get(ebRecHitsTag_)) {
+      if (passedEcalNoiseCut(hit, &thresholds)) {
+        TLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, iSetup.getData(caloGeometryToken_));
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+
+    for (const auto &hit : iEvent.get(eeRecHitsTag_)) {
+      if (passedEcalNoiseCut(hit, &thresholds)) {
+        TLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, iSetup.getData(caloGeometryToken_));
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+  }
+
+  bge_.set_particles(inputs);
+  iEvent.put(std::make_unique<double>(bge_.rho()));
+}
+
+void FixedGridRhoProducerFastjetFromRecHit::getHitP4(const DetId &detId,
+                                                     float hitE,
+                                                     TLorentzVector &hitp4,
+                                                     const CaloGeometry &caloGeometry) {
+  const CaloSubdetectorGeometry *subDetGeom = caloGeometry.getSubdetectorGeometry(detId);
+  std::shared_ptr<const CaloCellGeometry> cellGeom =
+      subDetGeom != nullptr ? subDetGeom->getGeometry(detId) : std::shared_ptr<const CaloCellGeometry>();
+  if (cellGeom != nullptr) {
+    const auto &gpPos = cellGeom->repPos();
```

Comment:
```suggestion
  if (subDetGeom != nullptr) {
    const auto &gpPos = subDetGeom->getGeometry(detId)->repPos();
```
This is a guess, just to highlight that maybe things could be simplified (..but maybe I'm missing something). Mainly, I didn't get why `shared_ptr` was needed.